File: trade-runner/inline-trade-executor.js
/**
 * 🎯 INLINE TRADE EXECUTOR - JavaScript verzija za n8n
 *
 * Stavlja se DIREKTNO nakon "IF: Trade Selected?" node-a.
 * Čita output od Trade Selector-a i izvršava trade automatski.
 *
 * KORISTI JavaScript Code node (ne Python!) jer samo JS može raditi HTTP pozive!
 *
 * ═══════════════════════════════════════════════════════════════
 */

// ═══════════════════════════════════════════════════════════════
// KONFIGURACIJA
// ═══════════════════════════════════════════════════════════════

// 🔑 STAVI SVOJE BYBIT API KREDENCIJALE OVDJE:
const BYBIT_API_KEY = "tvoj_api_key_ovdje";
const BYBIT_API_SECRET = "tvoj_api_secret_ovdje";

const BYBIT_API_URL = "https://api.bybit.com";
const MAX_LONG_POSITIONS = 5;
const MAX_SHORT_POSITIONS = 5;
const RECV_WINDOW = 5000;

// ═══════════════════════════════════════════════════════════════
// PURE JS HMAC SHA256 (bez require - radi u n8n sandbox!)
// ═══════════════════════════════════════════════════════════════

function sha256(message) {
    // Pure JavaScript SHA256 implementation
    function rotr(n, x) { return (x >>> n) | (x << (32 - n)); }
    function shr(n, x) { return x >>> n; }
    function ch(x, y, z) { return (x & y) ^ (~x & z); }
    function maj(x, y, z) { return (x & y) ^ (x & z) ^ (y & z); }
    function sigma0(x) { return rotr(2, x) ^ rotr(13, x) ^ rotr(22, x); }
    function sigma1(x) { return rotr(6, x) ^ rotr(11, x) ^ rotr(25, x); }
    function gamma0(x) { return rotr(7, x) ^ rotr(18, x) ^ shr(3, x); }
    function gamma1(x) { return rotr(17, x) ^ rotr(19, x) ^ shr(10, x); }

    const K = [
        0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1, 0x923f82a4, 0xab1c5ed5,
        0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3, 0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174,
        0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
        0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
        0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13, 0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
        0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
        0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
        0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208, 0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2
    ];

    // Convert string to UTF-8 bytes
    const utf8Encode = new TextEncoder();
    const msgBytes = utf8Encode.encode(message);
    const msgLen = msgBytes.length;
    const bitLen = msgLen * 8;

    // Padding
    const padLen = (msgLen % 64 < 56) ? (56 - msgLen % 64) : (120 - msgLen % 64);
    const totalLen = msgLen + padLen + 8;
    const padded = new Uint8Array(totalLen);
    padded.set(msgBytes);
    padded[msgLen] = 0x80;

    // Append length as 64-bit big-endian
    for (let i = 0; i < 8; i++) {
        padded[totalLen - 1 - i] = bitLen >>> (i * 8) & 0xff;
    }

    // Initialize hash values
    let H = [
        0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a,
        0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19
    ];

    // Process 512-bit chunks
    for (let chunk = 0; chunk < totalLen; chunk += 64) {
        const W = new Array(64);

        // Copy chunk into first 16 words
        for (let i = 0; i < 16; i++) {
            W[i] = (padded[chunk + i * 4] << 24) |
                   (padded[chunk + i * 4 + 1] << 16) |
                   (padded[chunk + i * 4 + 2] << 8) |
                   (padded[chunk + i * 4 + 3]);
        }

        // Extend to 64 words
        for (let i = 16; i < 64; i++) {
            W[i] = (gamma1(W[i - 2]) + W[i - 7] + gamma0(W[i - 15]) + W[i - 16]) >>> 0;
        }

        // Initialize working variables
        let [a, b, c, d, e, f, g, h] = H;

        // Main loop
        for (let i = 0; i < 64; i++) {
            const T1 = (h + sigma1(e) + ch(e, f, g) + K[i] + W[i]) >>> 0;
            const T2 = (sigma0(a) + maj(a, b, c)) >>> 0;
            h = g;
            g = f;
            f = e;
            e = (d + T1) >>> 0;
            d = c;
            c = b;
            b = a;
            a = (T1 + T2) >>> 0;
        }

        // Update hash values
        H[0] = (H[0] + a) >>> 0;
        H[1] = (H[1] + b) >>> 0;
        H[2] = (H[2] + c) >>> 0;
        H[3] = (H[3] + d) >>> 0;
        H[4] = (H[4] + e) >>> 0;
        H[5] = (H[5] + f) >>> 0;
        H[6] = (H[6] + g) >>> 0;
        H[7] = (H[7] + h) >>> 0;
    }

    // Produce final hash
    return H.map(h => h.toString(16).padStart(8, '0')).join('');
}

function hmacSha256(key, message) {
    const blockSize = 64; // SHA256 block size in bytes
    const utf8Encode = new TextEncoder();

    // Convert key to bytes
    let keyBytes = utf8Encode.encode(key);

    // Keys longer than blockSize are shortened
    if (keyBytes.length > blockSize) {
        keyBytes = new Uint8Array(sha256(key).match(/.{2}/g).map(byte => parseInt(byte, 16)));
    }

    // Keys shorter than blockSize are zero-padded
    if (keyBytes.length < blockSize) {
        const padded = new Uint8Array(blockSize);
        padded.set(keyBytes);
        keyBytes = padded;
    }

    // Compute inner and outer padded keys
    const oKeyPad = new Uint8Array(blockSize);
    const iKeyPad = new Uint8Array(blockSize);

    for (let i = 0; i < blockSize; i++) {
        oKeyPad[i] = keyBytes[i] ^ 0x5c;
        iKeyPad[i] = keyBytes[i] ^ 0x36;
    }

    // Convert to strings
    const oKeyPadStr = String.fromCharCode(...oKeyPad);
    const iKeyPadStr = String.fromCharCode(...iKeyPad);

    // Compute HMAC
    const innerHash = sha256(iKeyPadStr + message);
    const innerHashBytes = innerHash.match(/.{2}/g).map(byte => parseInt(byte, 16));
    const innerHashStr = String.fromCharCode(...innerHashBytes);

    return sha256(oKeyPadStr + innerHashStr);
}

function generateSignature(apiSecret, timestamp, recvWindow, paramStr) {
    return hmacSha256(apiSecret, paramStr);
}

async function bybitRequest(method, endpoint, params = null, body = null) {
    const timestamp = Date.now();
    let url = BYBIT_API_URL + endpoint;
    let requestBody = null;
    let paramStr;

    if (method === "GET") {
        const queryString = params
            ? Object.entries(params).map(([k, v]) => `${k}=${v}`).join('&')
            : '';
        paramStr = `${timestamp}${RECV_WINDOW}${queryString}`;
        if (queryString) url += `?${queryString}`;
    } else {
        const bodyJson = JSON.stringify(body || {});
        paramStr = `${timestamp}${RECV_WINDOW}${bodyJson}`;
        requestBody = bodyJson;
    }

    const signature = generateSignature(BYBIT_API_SECRET, timestamp, RECV_WINDOW, paramStr);

    const options = {
        method: method,
        url: url,
        headers: {
            'X-BAPI-API-KEY': BYBIT_API_KEY,
            'X-BAPI-TIMESTAMP': timestamp.toString(),
            'X-BAPI-SIGN': signature,
            'X-BAPI-RECV-WINDOW': RECV_WINDOW.toString(),
            'Content-Type': 'application/json'
        },
        body: requestBody,
        json: true
    };

    const response = await $httpRequest(options);
    return response;
}

// ═══════════════════════════════════════════════════════════════
// API FUNCTIONS
// ═══════════════════════════════════════════════════════════════

async function getPositions() {
    console.log("📊 Fetching positions...");
    const result = await bybitRequest('GET', '/v5/position/list', {
        category: 'linear',
        settleCoin: 'USDT'
    });

    if (result.retCode !== 0) {
        throw new Error(`Failed to get positions: ${result.retMsg}`);
    }

    const positions = result.result?.list || [];
    const activePositions = positions.filter(p => parseFloat(p.size || 0) > 0);
    console.log(`   Found ${activePositions.length} active positions`);
    return activePositions;
}

function validatePosition(symbol, side, positions) {
    const longCount = positions.filter(p => p.side === "Buy").length;
    const shortCount = positions.filter(p => p.side === "Sell").length;
    const existingSymbols = positions.map(p => p.symbol);

    console.log(`📊 Current: ${longCount} long / ${shortCount} short`);
    console.log(`   Symbols: ${existingSymbols.join(', ') || 'None'}`);
    console.log(`🔵 Request: ${symbol} ${side}`);

    if (existingSymbols.includes(symbol)) {
        return {
            approved: false,
            status: "skipped",
            reason: `⏭️ Position already exists for ${symbol}`,
            longCount,
            shortCount
        };
    }

    const isLong = side === "Buy";
    if (isLong && longCount >= MAX_LONG_POSITIONS) {
        return {
            approved: false,
            status: "rejected",
            reason: `🚫 Max long positions (${longCount}/${MAX_LONG_POSITIONS})`,
            longCount,
            shortCount
        };
    }

    if (!isLong && shortCount >= MAX_SHORT_POSITIONS) {
        return {
            approved: false,
            status: "rejected",
            reason: `🚫 Max short positions (${shortCount}/${MAX_SHORT_POSITIONS})`,
            longCount,
            shortCount
        };
    }

    console.log("✅ Position validation passed");
    return {
        approved: true,
        status: "approved",
        reason: "Position approved",
        longCount,
        shortCount
    };
}

async function setLeverage(symbol, leverage) {
    console.log(`⚙️ Setting leverage to ${leverage}x for ${symbol}`);
    const result = await bybitRequest('POST', '/v5/position/set-leverage', null, {
        category: 'linear',
        symbol: symbol,
        buyLeverage: leverage.toString(),
        sellLeverage: leverage.toString()
    });

    if (result.retCode !== 0) {
        throw new Error(`Failed to set leverage: ${result.retMsg}`);
    }

    console.log(`✅ Leverage set to ${leverage}x`);
    return result;
}

async function placeEntryOrder(symbol, side, qty, positionIdx = 0) {
    console.log(`🚀 Placing ${side} order for ${qty} ${symbol}`);
    const result = await bybitRequest('POST', '/v5/order/create', null, {
        category: 'linear',
        symbol: symbol,
        side: side,
        orderType: 'Market',
        qty: qty.toString(),
        positionIdx: positionIdx,
        timeInForce: 'IOC'
    });

    if (result.retCode !== 0) {
        throw new Error(`Failed to place entry order: ${result.retMsg}`);
    }

    const orderId = result.result?.orderId;
    console.log(`✅ Entry order placed: ${orderId}`);
    return result;
}

async function placeStopLoss(symbol, side, qty, stopLossPrice, positionIdx = 0) {
    console.log(`🛡️ Placing Stop Loss at ${stopLossPrice}`);
    const result = await bybitRequest('POST', '/v5/order/create', null, {
        category: 'linear',
        symbol: symbol,
        side: side,
        orderType: 'Market',
        qty: qty.toString(),
        stopLoss: stopLossPrice.toString(),
        triggerBy: 'LastPrice',
        reduceOnly: true,
        positionIdx: positionIdx
    });

    if (result.retCode !== 0) {
        throw new Error(`Failed to place stop loss: ${result.retMsg}`);
    }

    const orderId = result.result?.orderId;
    console.log(`✅ Stop Loss placed: ${orderId}`);
    return result;
}

async function placeTakeProfit(symbol, side, qty, takeProfitPrice, positionIdx = 0, label = "TP") {
    console.log(`💰 Placing ${label} at ${takeProfitPrice} for ${qty} ${symbol}`);
    const orderLinkId = `${label}_${symbol}_${Date.now()}`;

    const result = await bybitRequest('POST', '/v5/order/create', null, {
        category: 'linear',
        symbol: symbol,
        side: side,
        orderType: 'Limit',
        qty: qty.toString(),
        price: takeProfitPrice.toString(),
        reduceOnly: true,
        positionIdx: positionIdx,
        orderLinkId: orderLinkId
    });

    if (result.retCode !== 0) {
        throw new Error(`Failed to place ${label}: ${result.retMsg}`);
    }

    const orderId = result.result?.orderId;
    console.log(`✅ ${label} placed: ${orderId}`);
    return result;
}

// ═══════════════════════════════════════════════════════════════
// MAIN EXECUTION
// ═══════════════════════════════════════════════════════════════

async function executeTrade(tradeSignal) {
    try {
        const { symbol, side, qty, leverage, stop_loss, take_profits, position_idx, skip_leverage } = tradeSignal;

        console.log("\n" + "=".repeat(60));
        console.log("🎯 INLINE TRADE EXECUTOR - START");
        console.log("=".repeat(60));
        console.log(`Symbol: ${symbol}`);
        console.log(`Side: ${side}`);
        console.log(`Quantity: ${qty}`);
        console.log(`Leverage: ${leverage}x ${skip_leverage ? '(skip)' : ''}`);
        console.log(`Stop Loss: ${stop_loss}`);
        console.log(`Take Profits: ${take_profits?.length || 0}`);
        console.log("=".repeat(60) + "\n");

        // STEP 1: Get positions
        console.log("📊 STEP 1: Fetching current positions...");
        const positions = await getPositions();
        console.log();

        // STEP 2: Validate
        console.log("🔍 STEP 2: Validating position...");
        const validation = validatePosition(symbol, side, positions);
        console.log(`   Status: ${validation.status}\n`);

        if (!validation.approved) {
            console.log(`❌ Trade ${validation.status}: ${validation.reason}\n`);
            return {
                status: validation.status,
                symbol,
                side,
                qty,
                leverage,
                validation,
                reason: validation.reason
            };
        }

        // STEP 3: Set leverage
        if (skip_leverage) {
            console.log("⚙️ STEP 3: Skipping leverage (already set)\n");
        } else {
            console.log("⚙️ STEP 3: Setting leverage...");
            await setLeverage(symbol, leverage);
            console.log();
        }

        // STEP 4: Entry order
        console.log("🚀 STEP 4: Placing entry order...");
        const entryResult = await placeEntryOrder(symbol, side, qty, position_idx || 0);
        const entryOrderId = entryResult.result?.orderId;
        console.log();

        // STEP 5: Stop loss
        let slOrderId = null;
        if (stop_loss) {
            console.log("🛡️ STEP 5: Placing stop loss...");
            const oppositeSide = side === "Buy" ? "Sell" : "Buy";
            const slResult = await placeStopLoss(symbol, oppositeSide, qty, stop_loss, position_idx || 0);
            slOrderId = slResult.result?.orderId;
            console.log();
        }

        // STEP 6: Take profits
        const tpOrderIds = [];
        if (take_profits && take_profits.length > 0) {
            console.log(`💰 STEP 6: Placing ${take_profits.length} take profit orders...`);
            const oppositeSide = side === "Buy" ? "Sell" : "Buy";
            const totalQty = parseFloat(qty);

            for (const tp of take_profits) {
                const tpPrice = tp.price;
                const tpSizePct = tp.size_pct || 100;
                const tpLabel = tp.label || "TP";
                const tpQty = Math.floor(totalQty * (tpSizePct / 100));

                const tpResult = await placeTakeProfit(symbol, oppositeSide, tpQty, tpPrice, position_idx || 0, tpLabel);
                tpOrderIds.push(tpResult.result?.orderId);
            }
            console.log();
        }

        // SUCCESS
        console.log("=".repeat(60));
        console.log("✅ TRADE EXECUTION COMPLETED");
        console.log("=".repeat(60));
        console.log(`Entry Order: ${entryOrderId}`);
        console.log(`Stop Loss: ${slOrderId}`);
        console.log(`Take Profits: ${tpOrderIds.join(', ')}`);
        console.log("=".repeat(60) + "\n");

        return {
            status: "success",
            symbol,
            side,
            qty,
            leverage,
            entry_order_id: entryOrderId,
            sl_order_id: slOrderId,
            tp_order_ids: tpOrderIds,
            validation
        };

    } catch (error) {
        console.log(`\n❌ ERROR: ${error.message}\n`);
        return {
            status: "error",
            symbol: tradeSignal.symbol || "UNKNOWN",
            error: error.message
        };
    }
}

// ═══════════════════════════════════════════════════════════════
// n8n CODE NODE ENTRY POINT
// ═══════════════════════════════════════════════════════════════

(async () => {
    try {
        // Get input from Trade Selector
        const selectorOutput = items[0].json;

        console.log("\n🔄 TRANSFORMING TRADE SELECTOR OUTPUT...");
        console.log(`   Input: ${selectorOutput.symbol} ${selectorOutput.side}`);

        // Transform Trade Selector output → Trade Runner format
        // Side je već "Buy" ili "Sell" format - ne treba transformacija
        const tradePayload = {
            symbol: selectorOutput.symbol,
            side: selectorOutput.side,  // Već je "Buy" ili "Sell"
            qty: selectorOutput.qty,    // Već je string
            leverage: parseInt(selectorOutput.leverage),
            stop_loss: selectorOutput.stopLoss,  // camelCase!
            take_profits: selectorOutput.takeProfits || [],
            position_idx: selectorOutput.positionIdx || 0,
            skip_leverage: true  // Promijeniti na false ako želiš da postavlja leverage
        };

        console.log("✅ Transformation complete!");
        console.log(`   Symbol: ${tradePayload.symbol}`);
        console.log(`   Side: ${tradePayload.side}`);
        console.log(`   Qty: ${tradePayload.qty}`);
        console.log(`   Leverage: ${tradePayload.leverage}x\n`);

        // Execute trade
        const result = await executeTrade(tradePayload);

        // Return result
        return [{
            json: {
                ...result,
                original_selector_output: selectorOutput
            }
        }];

    } catch (error) {
        console.log(`❌ FATAL ERROR: ${error.message}`);
        console.log(error.stack);

        return [{
            json: {
                status: "error",
                error: error.message,
                stack: error.stack
            }
        }];
    }
})();

File: trade-runner/inline-trade-executor.py
"""
🎯 INLINE TRADE EXECUTOR - Direktno u workflow-u (bez webhook-a)

Stavlja se DIREKTNO nakon "IF: Trade Selected?" node-a.
Čita output od Trade Selector-a i izvršava trade automatski.

═══════════════════════════════════════════════════════════════
"""

import os
import hashlib
import hmac
import time
import json
import requests
from typing import Dict, List, Any, Optional

# ═══════════════════════════════════════════════════════════════
# KONFIGURACIJA
# ═══════════════════════════════════════════════════════════════

BYBIT_API_URL = "https://api.bybit.com"
MAX_LONG_POSITIONS = 5
MAX_SHORT_POSITIONS = 5
RECV_WINDOW = 5000

# ═══════════════════════════════════════════════════════════════
# BYBIT API FUNCTIONS (sve kao u Trade Runner-u)
# ═══════════════════════════════════════════════════════════════

def generate_signature(api_secret: str, timestamp: int, recv_window: int, query_string: str = "", body: str = "") -> str:
    param_str = str(timestamp) + str(recv_window) + (query_string or body)
    signature = hmac.new(
        api_secret.encode('utf-8'),
        param_str.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()
    return signature


def bybit_request(
    api_key: str,
    api_secret: str,
    method: str,
    endpoint: str,
    params: Optional[Dict] = None,
    body: Optional[Dict] = None
) -> Dict:
    timestamp = int(time.time() * 1000)
    url = BYBIT_API_URL + endpoint

    if method == "GET":
        query_string = "&".join([f"{k}={v}" for k, v in (params or {}).items()])
        signature = generate_signature(api_secret, timestamp, RECV_WINDOW, query_string=query_string)
        url += f"?{query_string}" if query_string else ""
        request_body = None
    else:
        body_json = json.dumps(body or {})
        signature = generate_signature(api_secret, timestamp, RECV_WINDOW, body=body_json)
        request_body = body_json

    headers = {
        "X-BAPI-API-KEY": api_key,
        "X-BAPI-TIMESTAMP": str(timestamp),
        "X-BAPI-SIGN": signature,
        "X-BAPI-RECV-WINDOW": str(RECV_WINDOW),
        "Content-Type": "application/json"
    }

    if method == "GET":
        response = requests.get(url, headers=headers, timeout=10)
    else:
        response = requests.post(url, headers=headers, data=request_body, timeout=10)

    response.raise_for_status()
    return response.json()


def get_positions(api_key: str, api_secret: str) -> List[Dict]:
    result = bybit_request(
        api_key=api_key,
        api_secret=api_secret,
        method="GET",
        endpoint="/v5/position/list",
        params={"category": "linear", "settleCoin": "USDT"}
    )
    if result.get("retCode") != 0:
        raise Exception(f"Failed to get positions: {result.get('retMsg')}")
    return result.get("result", {}).get("list", [])


def validate_position(symbol: str, side: str, positions: List[Dict]) -> Dict[str, Any]:
    active_positions = [p for p in positions if float(p.get("size", 0)) > 0]
    long_count = sum(1 for p in active_positions if p.get("side") == "Buy")
    short_count = sum(1 for p in active_positions if p.get("side") == "Sell")
    existing_symbols = [p.get("symbol") for p in active_positions]

    print(f"📊 Current Positions: {long_count} long / {short_count} short")
    print(f"   Symbols: {', '.join(existing_symbols) or 'None'}")
    print(f"🔵 Request: {symbol} {side}")

    if symbol in existing_symbols:
        return {
            "approved": False,
            "status": "skipped",
            "reason": f"⏭️ Position already exists for {symbol}",
            "long_count": long_count,
            "short_count": short_count
        }

    is_long = side == "Buy"
    if is_long and long_count >= MAX_LONG_POSITIONS:
        return {
            "approved": False,
            "status": "rejected",
            "reason": f"🚫 Max long positions ({long_count}/{MAX_LONG_POSITIONS})",
            "long_count": long_count,
            "short_count": short_count
        }

    if not is_long and short_count >= MAX_SHORT_POSITIONS:
        return {
            "approved": False,
            "status": "rejected",
            "reason": f"🚫 Max short positions ({short_count}/{MAX_SHORT_POSITIONS})",
            "long_count": long_count,
            "short_count": short_count
        }

    print("✅ Position validation passed")
    return {
        "approved": True,
        "status": "approved",
        "reason": "Position approved",
        "long_count": long_count,
        "short_count": short_count
    }


def set_leverage(api_key: str, api_secret: str, symbol: str, leverage: int) -> Dict:
    print(f"⚙️ Setting leverage to {leverage}x for {symbol}")
    result = bybit_request(
        api_key=api_key,
        api_secret=api_secret,
        method="POST",
        endpoint="/v5/position/set-leverage",
        body={
            "category": "linear",
            "symbol": symbol,
            "buyLeverage": str(leverage),
            "sellLeverage": str(leverage)
        }
    )
    if result.get("retCode") != 0:
        raise Exception(f"Failed to set leverage: {result.get('retMsg')}")
    print(f"✅ Leverage set to {leverage}x")
    return result


def place_entry_order(api_key: str, api_secret: str, symbol: str, side: str, qty: str, position_idx: int = 0) -> Dict:
    print(f"🚀 Placing {side} order for {qty} {symbol}")
    result = bybit_request(
        api_key=api_key,
        api_secret=api_secret,
        method="POST",
        endpoint="/v5/order/create",
        body={
            "category": "linear",
            "symbol": symbol,
            "side": side,
            "orderType": "Market",
            "qty": str(qty),
            "positionIdx": position_idx,
            "timeInForce": "IOC"
        }
    )
    if result.get("retCode") != 0:
        raise Exception(f"Failed to place entry order: {result.get('retMsg')}")
    order_id = result.get("result", {}).get("orderId")
    print(f"✅ Entry order placed: {order_id}")
    return result


def place_stop_loss(api_key: str, api_secret: str, symbol: str, side: str, qty: str, stop_loss_price: str, position_idx: int = 0) -> Dict:
    print(f"🛡️ Placing Stop Loss at {stop_loss_price}")
    result = bybit_request(
        api_key=api_key,
        api_secret=api_secret,
        method="POST",
        endpoint="/v5/order/create",
        body={
            "category": "linear",
            "symbol": symbol,
            "side": side,
            "orderType": "Market",
            "qty": str(qty),
            "stopLoss": str(stop_loss_price),
            "triggerBy": "LastPrice",
            "reduceOnly": True,
            "positionIdx": position_idx
        }
    )
    if result.get("retCode") != 0:
        raise Exception(f"Failed to place stop loss: {result.get('retMsg')}")
    order_id = result.get("result", {}).get("orderId")
    print(f"✅ Stop Loss placed: {order_id}")
    return result


def place_take_profit(api_key: str, api_secret: str, symbol: str, side: str, qty: str, take_profit_price: str, position_idx: int = 0, label: str = "TP") -> Dict:
    print(f"💰 Placing {label} at {take_profit_price} for {qty} {symbol}")
    order_link_id = f"{label}_{symbol}_{int(time.time() * 1000)}"
    result = bybit_request(
        api_key=api_key,
        api_secret=api_secret,
        method="POST",
        endpoint="/v5/order/create",
        body={
            "category": "linear",
            "symbol": symbol,
            "side": side,
            "orderType": "Limit",
            "qty": str(qty),
            "price": str(take_profit_price),
            "reduceOnly": True,
            "positionIdx": position_idx,
            "orderLinkId": order_link_id
        }
    )
    if result.get("retCode") != 0:
        raise Exception(f"Failed to place {label}: {result.get('retMsg')}")
    order_id = result.get("result", {}).get("orderId")
    print(f"✅ {label} placed: {order_id}")
    return result


def execute_trade(trade_signal: Dict) -> Dict:
    try:
        api_key = trade_signal.get("api_key")
        api_secret = trade_signal.get("api_secret")
        symbol = trade_signal.get("symbol")
        side = trade_signal.get("side")
        qty = str(trade_signal.get("qty"))
        leverage = int(trade_signal.get("leverage", 1))
        stop_loss = trade_signal.get("stop_loss")
        take_profits = trade_signal.get("take_profits", [])
        position_idx = int(trade_signal.get("position_idx", 0))
        skip_leverage = trade_signal.get("skip_leverage", False)

        print("\n" + "=" * 60)
        print("🎯 INLINE TRADE EXECUTOR - START")
        print("=" * 60)
        print(f"Symbol: {symbol}")
        print(f"Side: {side}")
        print(f"Quantity: {qty}")
        print(f"Leverage: {leverage}x {'(skip)' if skip_leverage else ''}")
        print(f"Stop Loss: {stop_loss}")
        print(f"Take Profits: {len(take_profits)}")
        print("=" * 60 + "\n")

        # STEP 1: Get current positions
        print("📊 STEP 1: Fetching current positions...")
        positions = get_positions(api_key, api_secret)
        print(f"   Found {len(positions)} positions\n")

        # STEP 2: Validate position
        print("🔍 STEP 2: Validating position...")
        validation = validate_position(symbol, side, positions)
        print(f"   Status: {validation['status']}\n")

        if not validation["approved"]:
            print(f"❌ Trade {validation['status']}: {validation['reason']}\n")
            return {
                "status": validation["status"],
                "symbol": symbol,
                "side": side,
                "qty": qty,
                "leverage": leverage,
                "validation": validation,
                "reason": validation["reason"]
            }

        # STEP 3: Set leverage (optional)
        if skip_leverage:
            print("⚙️ STEP 3: Skipping leverage (already set)\n")
        else:
            print("⚙️ STEP 3: Setting leverage...")
            set_leverage(api_key, api_secret, symbol, leverage)
            print()

        # STEP 4: Place entry order
        print("🚀 STEP 4: Placing entry order...")
        entry_result = place_entry_order(api_key, api_secret, symbol, side, qty, position_idx)
        entry_order_id = entry_result.get("result", {}).get("orderId")
        print()

        # STEP 5: Place stop loss
        sl_order_id = None
        if stop_loss:
            print("🛡️ STEP 5: Placing stop loss...")
            opposite_side = "Sell" if side == "Buy" else "Buy"
            sl_result = place_stop_loss(api_key, api_secret, symbol, opposite_side, qty, stop_loss, position_idx)
            sl_order_id = sl_result.get("result", {}).get("orderId")
            print()

        # STEP 6: Place take profits
        tp_order_ids = []
        if take_profits:
            print(f"💰 STEP 6: Placing {len(take_profits)} take profit orders...")
            opposite_side = "Sell" if side == "Buy" else "Buy"
            total_qty = float(qty)

            for tp in take_profits:
                tp_price = tp.get("price")
                tp_size_pct = tp.get("size_pct", 100)
                tp_label = tp.get("label", "TP")
                tp_qty = int(total_qty * (tp_size_pct / 100))

                tp_result = place_take_profit(api_key, api_secret, symbol, opposite_side, str(tp_qty), tp_price, position_idx, tp_label)
                tp_order_id = tp_result.get("result", {}).get("orderId")
                tp_order_ids.append(tp_order_id)
            print()

        # SUCCESS
        print("=" * 60)
        print("✅ TRADE EXECUTION COMPLETED")
        print("=" * 60)
        print(f"Entry Order: {entry_order_id}")
        print(f"Stop Loss: {sl_order_id}")
        print(f"Take Profits: {', '.join(tp_order_ids)}")
        print("=" * 60 + "\n")

        return {
            "status": "success",
            "symbol": symbol,
            "side": side,
            "qty": qty,
            "leverage": leverage,
            "entry_order_id": entry_order_id,
            "sl_order_id": sl_order_id,
            "tp_order_ids": tp_order_ids,
            "validation": validation
        }

    except Exception as e:
        print(f"\n❌ ERROR: {str(e)}\n")
        return {
            "status": "error",
            "symbol": trade_signal.get("symbol", "UNKNOWN"),
            "error": str(e)
        }


# ═══════════════════════════════════════════════════════════════
# n8n CODE NODE ENTRY POINT - TRANSFORM + EXECUTE
# ═══════════════════════════════════════════════════════════════

try:
    # Get input from Trade Selector
    selector_output = $input.item.json

    print("\n🔄 TRANSFORMING TRADE SELECTOR OUTPUT...")
    print(f"   Input: {selector_output.get('symbol')} {selector_output.get('side')}")

    # Transform Trade Selector output → Trade Runner format
    bybit_side = "Buy" if selector_output.get("side") == "BUY" else "Sell"

    trade_payload = {
        "symbol": selector_output.get("bybit_symbol") or selector_output.get("symbol"),
        "side": bybit_side,
        "qty": str(selector_output.get("quantity")),
        "leverage": selector_output.get("leverage"),
        "stop_loss": str(selector_output.get("stop_loss")),
        "take_profits": selector_output.get("take_profits", []),
        "position_idx": 0,
        "skip_leverage": True,  # Promijeniti na False ako želiš da postavlja leverage

        # API credentials from environment
        "api_key": os.getenv('BYBIT_API_KEY') or os.getenv('bybit_api_key'),
        "api_secret": os.getenv('BYBIT_API_SECRET') or os.getenv('bybit_api_secret')
    }

    print("✅ Transformation complete!")
    print(f"   Symbol: {trade_payload['symbol']}")
    print(f"   Side: {trade_payload['side']}")
    print(f"   Qty: {trade_payload['qty']}")
    print(f"   Leverage: {trade_payload['leverage']}x\n")

    # Execute trade
    result = execute_trade(trade_payload)

    # Return result
    return [{
        "json": {
            **result,
            "original_selector_output": selector_output
        }
    }]

except Exception as e:
    print(f"❌ FATAL ERROR: {str(e)}")
    import traceback
    traceback.print_exc()

    return [{
        "json": {
            "status": "error",
            "error": str(e),
            "traceback": traceback.format_exc()
        }
    }]
